[SPARK-47336][SQL][CONNECT] Provide to PySpark a functionality to get estimated size of DataFrame in bytes

SemyonSinchenko opened the pull request:
### What changes were proposed in this pull request?

In PySpark connect there is no access to JVM to call `queryExecution().optimizedPlan.stats`. So, there is no way to get information about size in bytes from plan except parsing by regexps an output of `explain`. This PR is trying to fill that gap by providing `sizeInBytesApproximation` method to JVM, PySpark Classic and PySpark Connect APIs. Under the hood it is just a call to `queryExecution().optimizedPlan.stats.sizeInBytes`. JVM and PySpark Classic APIs were updated just to have a parity.

1. Update of `Dataset.scala` in JVM connect by adding a new API
2. Update of `Dataset.scala` in JVM classic by adding a new API
3. Update `dataframe.py` in sql by adding signature and doc of a new API
4. Update `dataframe.py` in connect by adding an implementation of a new API
5. Update `dataframe.py` in classic by adding an implementation of a new API
6. Update `base.proto` in part `AnalyzeRequest` / `AnalyzeResponse` by adding new message
7. Generate new py-files from proto
8. Update `SparkConnectAnalyzeHandler` by extending `match` and adding call to `queryExecution`
9. Update `SparkConnectClient` by adding a new method that build a new request
10. Update `SparkSession` by adding a call to client and parsing a response
11. Add/update corresponding tests

### Why are the changes needed?
To provide to PySpark Connect users an ability to get in runtime the DataFrame size estimation without forcing them to parse string-output of `df.explain`. Other changes are needed to have a parity across Connect / Classic and PySpark / JVM Spark.

### Does this PR introduce _any_ user-facing change?
Only a new API. The new API is mostly for PySpark Connect users.

### How was this patch tested?
Because the actual logic is in `queryExecution` I added tests only for syntax / calls. In tests we are testing that for a dataframe the returned size is greater than zero.

### Was this patch authored or co-authored using generative AI tooling?
No.


@grundprinzip We discussed that ticket with you, may you please make a look? Thanks!

SemyonSinchenko committed:
Add sizeInBytes to Connect and Dataset

 On branch size_in_bytes_api
 Changes to be committed:
	modified:   connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala
	modified:   connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala
	modified:   connector/connect/common/src/main/protobuf/spark/connect/base.proto
	modified:   connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAnalyzeHandler.scala
	modified:   sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

SemyonSinchenko committed:
Fix scalastyle problems + update proto

 Changes to be committed:
	modified:   .gitignore
	modified:   connector/connect/common/src/main/protobuf/spark/connect/base.proto
	modified:   connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAnalyzeHandler.scala
	modified:   sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

SemyonSinchenko committed:
Fixes

+ Relation instead of Plan in base.proto
+ Fix broken ids in base.proto
+ Fix corresponding parts in AnalyzeHandler

 On branch size_in_bytes_api
 Your branch is up to date with 'origin/size_in_bytes_api'.

 Changes to be committed:
	modified:   connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala
	modified:   connector/connect/common/src/main/protobuf/spark/connect/base.proto
	modified:   connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAnalyzeHandler.scala

SemyonSinchenko committed:
Make API works

+ update naming following the discussion in JIRA

 On branch size_in_bytes_api
 Your branch is up to date with 'origin/size_in_bytes_api'.

 Changes to be committed:
	modified:   connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala
	modified:   connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala
	modified:   connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala
	modified:   connector/connect/common/src/main/protobuf/spark/connect/base.proto
	modified:   connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala
	modified:   connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAnalyzeHandler.scala

SemyonSinchenko committed:
Update Python connect and classic APIs

+ small fixes
+ tests

 On branch size_in_bytes_api
 Your branch is up to date with 'origin/size_in_bytes_api'.

 Changes to be committed:
	modified:   python/pyspark/sql/classic/dataframe.py
	modified:   python/pyspark/sql/connect/client/core.py
	modified:   python/pyspark/sql/connect/dataframe.py
	modified:   python/pyspark/sql/connect/proto/base_pb2.py
	modified:   python/pyspark/sql/connect/proto/base_pb2.pyi
	modified:   python/pyspark/sql/dataframe.py
	modified:   python/pyspark/sql/tests/connect/test_connect_basic.py
	modified:   python/pyspark/sql/tests/test_dataframe.py
	modified:   sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

SemyonSinchenko committed:
Docstring example and sync connect/classic jvm

 On branch size_in_bytes_api
 Your branch is up to date with 'origin/size_in_bytes_api'.

 Changes to be committed:
	modified:   python/pyspark/sql/dataframe.py
	modified:   sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

SemyonSinchenko committed:
Fix linter

+ delete an example because it requires a data

 On branch size_in_bytes_api
 Your branch is up to date with 'origin/size_in_bytes_api'.

 Changes to be committed:
	modified:   python/pyspark/sql/connect/dataframe.py
	modified:   python/pyspark/sql/dataframe.py

SemyonSinchenko committed:
Fix MyPy

 On branch size_in_bytes_api
 Your branch is up to date with 'origin/size_in_bytes_api'.

 Changes to be committed:
	modified:   python/pyspark/sql/connect/dataframe.py

HyukjinKwon commented on .gitignore:
```diff
@@ -26,6 +26,7 @@
 .scala_dependencies
 .settings
 .vscode
+.dir-locals.el
```
let's remove this

SemyonSinchenko replied:
Done

HyukjinKwon commented on python/pyspark/sql/connect/client/core.py:
```diff
@@ -1157,6 +1163,20 @@ def _analyze_plan_request_with_metadata(self) -> pb2.AnalyzePlanRequest:
             req.user_context.user_id = self._user_id
         return req
 
+    def _sizeInBytes(self, relation: pb2.Relation) -> AnalyzeResult:
```
`_size_in_bytes`

zhengruifeng commented on connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -283,6 +283,16 @@ class Dataset[T] private[sql] (
+   * @group basic
+   * @since 4.0.0
+   */
+  def sizeInBytesApproximation(): Long = {
```
It seems the size may exceed `Long.MaxValue`:

https://github.com/apache/spark/blob/111529e0dea68bf5343cc6aabba53b59e5d21830/core/src/main/scala/org/apache/spark/util/Utils.scala#L1147-L1159

SemyonSinchenko replied:
Thank you! I missed that. I changed the type to BigInteger (bytes[] in proto)

SemyonSinchenko committed:
Updates from comments

- change from Long to bytes[] in proto
- JVM methods return BigInteger from now
- in Python conversion from BigInteger to int is via bytes[]
- drop .dir-locals.el from .gitignore
- rename _sizeInBytes -> _size_in_bytes on Python side

 On branch size_in_bytes_api
 Your branch is up to date with 'origin/size_in_bytes_api'.

 Changes to be committed:
	modified:   .gitignore
	modified:   connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala
	modified:   connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala
	modified:   connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala
	modified:   connector/connect/common/src/main/protobuf/spark/connect/base.proto
	modified:   connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAnalyzeHandler.scala
	modified:   python/pyspark/sql/classic/dataframe.py
	modified:   python/pyspark/sql/connect/client/core.py
	modified:   python/pyspark/sql/connect/dataframe.py
	modified:   python/pyspark/sql/connect/proto/base_pb2.py
	modified:   python/pyspark/sql/connect/proto/base_pb2.pyi
	modified:   python/pyspark/sql/dataframe.py
	modified:   sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

SemyonSinchenko committed:
Fix scalastyle

 On branch size_in_bytes_api
 Your branch is ahead of 'origin/size_in_bytes_api' by 1 commit.
   (use "git push" to publish your local commits)

 Changes to be committed:
	modified:   connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala
	modified:   connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala
	modified:   connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAnalyzeHandler.scala

SemyonSinchenko committed:
Fix scalastyle part2

 On branch size_in_bytes_api
 Your branch is ahead of 'origin/size_in_bytes_api' by 2 commits.
   (use "git push" to publish your local commits)

 Changes to be committed:
	modified:   sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

SemyonSinchenko committed:
Scalastyle part-3

 On branch size_in_bytes_api
 Your branch is ahead of 'origin/size_in_bytes_api' by 3 commits.
   (use "git push" to publish your local commits)

 Changes to be committed:
	modified:   connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala

SemyonSinchenko committed:
Fix py-linter

 On branch size_in_bytes_api
 Your branch is up to date with 'origin/size_in_bytes_api'.

 Changes to be committed:
	modified:   python/pyspark/sql/classic/dataframe.py

SemyonSinchenko commented on python/pyspark/sql/dataframe.py:
```diff
@@ -657,6 +657,19 @@ def printSchema(self, level: Optional[int] = None) -> None:
+        Returns
+        -------
+        int
+            Size in bytes estimated from optimized plan
```
Done

SemyonSinchenko commented:
New changes:
- fixes from comments
- **changing the type from Long to BigInteger** (`bytes` in proto)

SemyonSinchenko committed:
Merge branch 'master' into size_in_bytes_api

+ resolving conflicts
+ regenerate python proto-classes

SemyonSinchenko commented:
@HyukjinKwon sorry for tagging, but may you please make a look again? Thanks in advance!

SemyonSinchenko committed:
Merge branch 'master' into size_in_bytes_api

SemyonSinchenko committed:
Re-generate proto files and update docstring

I updated docstring for sizeInBytes method of dataframe.

 Changes to be committed:
	modified:   python/pyspark/sql/connect/proto/base_pb2.py
	modified:   python/pyspark/sql/dataframe.py

SemyonSinchenko commented:
Changes from the last two commits (actual changes marked by bold):
- resolve merge conflicts
- re-generate proto files for PySpark
- **update docstring in** `dataframe.py`**: fix a typo and extend it by describing how it works and corner-cases**

SemyonSinchenko committed:
Fix scalafmt

 Changes to be committed:
	modified:   connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala
	modified:   connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala

SemyonSinchenko committed:
Fix pylint

SemyonSinchenko commented:
@HyukjinKwon I'm sorry for tagging you again, but maybe you can make a look? Thanks in advance!

SemyonSinchenko committed:
Merge branch 'master' into size_in_bytes_api

SemyonSinchenko commented:
@HyukjinKwon @zhengruifeng Sorry for tagging but maybe you can take a look again? I fixed everything from the last review round... Tnx in advance!

SemyonSinchenko committed:
Merge remote-tracking branch 'apache/master' into size_in_bytes_api

SemyonSinchenko committed:
Merge remote-tracking branch 'apache/master' into size_in_bytes_api

SemyonSinchenko committed:
Merge remote-tracking branch 'apache/master' into size_in_bytes_api

SemyonSinchenko commented:
@HyukjinKwon @zhengruifeng Sorry for tagging but maybe you can take a look again? I fixed everything from the last review round... Tnx in advance!

cloud-fan commented:
@SemyonSinchenko thanks for the contribution! I have two high-level questions:

- How do you use this sizeInBytes? Is it accurate in your workload?
- Shall we expose more stats like numRows?

SemyonSinchenko commented:
> @SemyonSinchenko thanks for the contribution! I have two high-level questions:
> 
> * How do you use this sizeInBytes? Is it accurate in your workload?
> * Shall we expose more stats like numRows?

@cloud-fan Thank you for the comment!

The main usage of that is for library developers / developers of the reusable spark code. If one wants to use, for example, a broadcast hint inside the library code / reusable code without changing a thresholds globally the only way to get an estimated size of the data is from the plan. But in Connect there is no way to call `queryExecution().optimizedPlan.stats` because there is no JavaBridge. So, at the moment, Connect devs can get this information only from the parsing of the string-representation of the plan via regexps. But this way is very unstable and fragile to versions update, because obviously there are zero guarantees that the string representation of the spark plan won't change in the future versions. This estimation is very inaccurate because it is an upper bound. But upper bound is perfect for that estimation, because devs has a guarantee that the size of the data is not more than the output of that function.

As an example I can point to an implementation of one of the databricks libs, named tempo: [code-example](https://github.com/databrickslabs/tempo/blob/master/python/tempo/tsdf.py#L712). But it is only an example: I have no relation to databricks, I'm not a contributor of tempo. We are using the similar techniques internally at the moment.


More stats can be exposed, but actually I do not see usage of the `numRows` estimation, for example. The most usable is size in bytes estimation because it can be used for an estimation of the possibility of the `collect`, broadcast hints, etc.

SemyonSinchenko committed:
Merge remote-tracking branch 'apache/master' into size_in_bytes_api

SemyonSinchenko committed:
Merge remote-tracking branch 'apache/master' into size_in_bytes_api

SemyonSinchenko committed:
Merge remote-tracking branch 'apache/master' into size_in_bytes_api

SemyonSinchenko commented:
@HyukjinKwon @zhengruifeng @cloud-fan Sorry for tagging but maybe you can take a look again? I fixed everything from the last review round... Tnx in advance!

SemyonSinchenko committed:
Merge remote-tracking branch 'apache/master' into size_in_bytes_api

SemyonSinchenko committed:
Fix scalastyle

SemyonSinchenko committed:
Merge remote-tracking branch 'apache/master' into size_in_bytes_api

SemyonSinchenko commented:
@HyukjinKwon @zhengruifeng @cloud-fan Sorry for tagging but maybe you can take a look again? I fixed everything from the last review round... Tnx in advance!

SemyonSinchenko committed:
Merge remote-tracking branch 'apache/master'

SemyonSinchenko commented:
@HyukjinKwon @zhengruifeng @cloud-fan Sorry for tagging but maybe you can take a look again? I fixed everything from the last review round... Tnx in advance!

zhengruifeng commented:
@SemyonSinchenko My major concern is whether we need this API:
IMHO, the estimated size is not precise enough to be added as a dataset/dataframe API, I view `queryExecution().optimizedPlan.stats.sizeInBytes` as kind of developer API only for advanced users and developers.

zhengruifeng commented:
@SemyonSinchenko [ExecutionInfo](https://github.com/apache/spark/commit/9d4abaf19fa4d508aca35ac00528b2ce9f4e8805) was recently add to expose some query execution metrics.

Not sure whether it is the right place to add the estimated size. cc @grundprinzip
